### PR TITLE
Reimplement (Fast)ResizeCropMirror in terms of Resize. Remove dead code.

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**Refactoring**
**Bug fix**
**Breaking change**
**New feature**

## Description:
Recatoring:
This PR refactors the rarely used ResizeCropMirror operator and implements it in terms of Resize.

New feature:
ResizeCropMirror has GPU backend now, for free.
ResizeCropMirror actually looks at `"dtype"` argument.
ResizeCropMirror has all the goodies of Resize and RandomResizedCrop - antialiasing, different resampling modes, resize modes ("not_larger", "not_smaller", etc), specifying size as a single vector argument `"size"`.

Bug fix which is also a breaking change:
Some parameters, which were accidentally added to ResizeCropMirror, are no longer accepted

`dtype` handling
The old operator had a `dtype` argument that could take float, float16 or uint8, but was in fact completely ignored.
Now it has the same restrictions as `dtype` in `Resize`.

Breaking change:
By default, ResizeCropMirror applies antialiasing (as every other resize flavor); I kept it disabled in FastResizeCropMirror, though.

The use code that specified the ignored arguments will become invalid, but it was already not performing the things that the documentation claimed.

## Additional information:
The old variant used a custom implementation of resizing. There were two flavors:
- one which calculated warp parameters an used warping
- one which just resized and cropped the image
Both of them are removed.

The old implementation had a bug in Schema - it inherited Crop instead of CropAttr, and thus had a bunch of arguments which were accepted, but completely ignored by the operator; this has been fixed.

### Affected modules and functionalities:
ResizeAttr
(Fast)ResizeCropMirror

### Key points relevant for the review:
ROI fixup code?

### Tests:
The tests have been modified to include the GPU variant and the *Warp variant has been removed, because there's just one code path now.
I've also added Python tests vs separate resize, crop, flip operators.

- [X] Existing tests apply
- [X] New tests added
  - [X] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [X] Documentation updated
  - [X] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-737

### CODE REVIEW

File: dali/operators/image/resize/resize_attr.cc [hunk 1]
```diff
@@ -213,45 +239,21 @@ void ResizeAttr::PrepareResizeParams(const OpSpec &spec, const ArgumentWorkspace
...
     for (int i = 0; i < N; i++) {
-      for (int d = 0; d < spatial_ndim_; d++) {
-        requested_size[d] = size_vecs[d][i];
-      }
-
-      bool empty_input = volume(input_shape.tensor_shape_span(i)) == 0;
-      CalculateInputRoI(in_lo, in_hi, has_roi_, roi_relative_, make_cspan(roi_start_),
-                        make_cspan(roi_end_), input_shape, i, spatial_ndim_, first_spatial_dim_);
-      CalculateSampleParams(params_[i], requested_size, in_lo, in_hi, subpixel_scale_, empty_input,
-                            spatial_ndim_, mode_, max_size);
```
Comment (mzient) [inline dali/operators/image/resize/resize_attr.cc:224]: I initially misunderstood how ResizeCropMirror works, so I made some customization points here. There was a lot of hastily written repetitive code which I refactored in the process. After discovering my mistake, I decided to keep those changes.

File: dali/operators/image/resize/resize_crop_mirror.cc [hunk 2]
```diff
@@ -16,41 +16,147 @@
...
+  .InputLayout(0, {"HWC",  "FHWC",  "CHW",  "FCHW",  "CFHW" ,
+                   "DHWC", "FDHWC", "CDHW", "FCDHW", "CFDHW"  });
 
 DALI_SCHEMA(FastResizeCropMirror)
-  .DocStr(R"code(Performs a fused resize, crop, mirror operation.
-
-The operator handles both fixed and random resizing and cropping.
-Backprojects the desired crop through the resize operation to reduce the amount of
-work that is performed.)code")
+  .DocStr(R"(Legacy alias for ResizedCropMirror, with antialiasing disabled by defaul.)")
```
Comment (jantonguirao) [inline dali/operators/image/resize/resize_crop_mirror.cc:49]:
```suggestion
  .DocStr(R"(Legacy alias for ResizedCropMirror, with antialiasing disabled by default.)")
```

**Comment by klecki** [comment]
I looked only into the generated docs to compare the APIs. 
Old `dtype`:
> Supported types: FLOAT, FLOAT16, and UINT8.
New `dtype`:
> Must be same as input type or float. If not set, input type is used.
Is this a breaking change?

The crop parameters: `fill_values` and `out_of_bounds_policy` were removed. Were they previously supported, or did they cause no-op or error?

`image_type` and `output_dtype` - we wanted to have the deprecation policy after N releases, but we didn't do it. Should we remove them now? There are other places in DALI with such parameters.

We still have the deprecation option, which basically keeps an argument in schema, but ignores it automatically. I think, as this was mostly a legacy operator, we should consider keeping the APIs stable.

**Comment by mzient** [comment]
> I looked only into the generated docs to compare the APIs. Old `dtype`:
> 
> > Supported types: FLOAT, FLOAT16, and UINT8.
> > New `dtype`:
> > Must be same as input type or float. If not set, input type is used.
> > Is this a breaking change?
It's a fixing change. This argument was ignored. The only supported type (input and output) was uint8.

> The crop parameters: `fill_values` and `out_of_bounds_policy` were removed. Were they previously supported, or did they cause no-op or error?

They were ignored. I decided to do something about this op largely because someone complained on GH about these parameters not having any effect. They were added by mistake - there was `AddParent("Crop")` instead of "CropAttr"`.

> `image_type` and `output_dtype` - we wanted to have the deprecation policy after N releases, but we didn't do it. Should we remove them now? There are other places in DALI with such parameters.

I think we can remove them now without looking back much; this operator was barely functional anyway. For what it's worth, `image_type` is still there, inherited from `resize`.

> 
> We still have the deprecation option, which basically keeps an argument in schema, but ignores it automatically. I think, as this was mostly a legacy operator, we should consider keeping the APIs stable.

I'd rather fix it and keep it _functional_. It wasn't.

**Review by jantonguirao — APPROVED** [review]

File: dali/test/python/operator_2/test_resize_crop_mirror.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    crop_h = fn.random.uniform(range=(100, 224), dtype=types.FLOAT)
+    crop_x = fn.random.uniform(range=(0, 1))
+    crop_y = fn.random.uniform(range=(0, 1))
+    out = fn.resize_crop_mirror(images, size=size, mode=mode,
+                                crop_w=crop_w, crop_h=crop_h, crop_pos_x=crop_x, crop_pos_y=crop_y,
+                                mirror=flip)
+    resized = fn.resize(images, size=size, mode=mode)
+    cropped = fn.crop(resized, crop_w=crop_w, crop_h=crop_h, crop_pos_x=crop_x, crop_pos_y=crop_y)
+    flipped = fn.flip(cropped, horizontal=flip_x, vertical=flip_y)
+    return out, flipped
```
Comment (klecki) [inline dali/test/python/operator_2/test_resize_crop_mirror.py:46]: Can you test the mix of roi for resize & crop, (most notably the one that you mentioned, where the roi in resize already flips)?

**Review by klecki — COMMENTED** [review]
Looks ok, just GSG issues + ask for additional test.

File: dali/operators/image/resize/resize_crop_mirror.cc [hunk 4]
```diff
@@ -16,41 +16,149 @@
...
+      false)
+  .InputLayout(0, {"HWC",  "FHWC",  "CHW",  "FCHW",  "CFHW" ,
+                   "DHWC", "FDHWC", "CDHW", "FCDHW", "CFDHW"  });
+
+template<typename Backend>
+ResizeCropMirror<Backend>::ResizeCropMirror(const OpSpec &spec)
+    : StatelessOperator<Backend>(spec)
+    , ResizeBase<Backend>(spec)
+    , resize_attr_(spec) {
+  resample_params_.resize(num_threads_);
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/operators/image/resize/resize_crop_mirror.cc:70]: Probably...

File: dali/test/python/operator_2/test_resize_crop_mirror.py (hunk 3 above)
Comment (mzient) [reply dali/test/python/operator_2/test_resize_crop_mirror.py:46]: Why not.

Comment (mzient) [reply dali/test/python/operator_2/test_resize_crop_mirror.py]: Done.

File: dali/operators/image/resize/resize_crop_mirror.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (mzient) [reply dali/operators/image/resize/resize_crop_mirror.cc]: Yup, removed and nothing bad happened.

File: dali/operators/image/resize/resize_crop_mirror.cc [hunk 5]
```diff
@@ -16,41 +16,148 @@
...
+  Nearest neighbor interpolation does not support antialiasing.)",
+      false)
+  .InputLayout(0, {"HWC",  "FHWC",  "CHW",  "FCHW",  "CFHW" ,
+                   "DHWC", "FDHWC", "CDHW", "FCDHW", "CFDHW"  });
+
+template<typename Backend>
+ResizeCropMirror<Backend>::ResizeCropMirror(const OpSpec &spec)
+    : StatelessOperator<Backend>(spec)
+    , ResizeBase<Backend>(spec)
+    , resize_attr_(spec) {
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/operators/image/resize/resize_crop_mirror.cc:69]: It's not that uncommon - there are hundreds of lines in our code that start with a comma and you can find such formatting in other places, too (cutlass, cv-cuda).
Also, the consensus was that formatting suggestions that are beyond what linter requires are not not something over which we would block PRs.

**Review by klecki — APPROVED** [review]
Changes approved, the styling not. Let's bring it up during the next meeting.

File: dali/operators/image/resize/resize_crop_mirror.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/image/resize/resize_crop_mirror.cc:69]: Well, we state in https://github.com/NVIDIA/DALI/blob/main/CONTRIBUTING.md:
```
DALI Coding Style Guide can be found [here](https://github.com/NVIDIA/DALI/blob/main/STYLE_GUIDE.md). We follow Google C++ Style Guide with few exceptions and additional guidelines regarding DALI-specific cases. See the DALI Coding Style Guide for details. When no rules can be found, follow the already occuring conventions. If there is no precedence in our codebase we are open to discussion.```
So this is confusing to external contributors who can be surprised by us not sticking to our own rules (not to mention several cases like https://github.com/NVIDIA/DALI/pull/3982 where we put a lot of effort not just make the linter happy but also make code consistent and readable).

**Pull request merged**